src/lib/gossip_net/gossip_net.ml

```diff
@@ -239,25 +240,40 @@ module Make (Message : Message_intf) :
                     "%s" (Exn.to_string_mach exn) ))
             (Tcp.Where_to_listen.of_port config.me.Peer.communication_port)
             (fun client reader writer ->
-              let%map _ =
-                Rpc.Connection.server_with_close reader writer ~implementations
-                  ~connection_state:(fun conn ->
-                    (* connection state is the client's IP and ephemeral port
+              let conn_list =
+                Option.value_map ~default:[]
+                  (Hashtbl.find t.connections (Socket.Address.Inet.addr client))
+                  ~f:Fn.id
+              in
+              let connection_limit = List.length implementation_list + 1 in
+              if List.length conn_list = connection_limit then (
+                Logger.error t.logger ~module_:__MODULE__ ~location:__LOC__
+                  "Cannot open another connection. Number of existing \
+                   connections from %s equals the limit %d."
+                  (Socket.Address.Inet.to_string client)
+                  connection_limit ;
+                Deferred.unit )
```

Comment:
I think we should be closing the connection explicitly here.

Reply:
You mean all the existing connections?

Reply:
> We shouldn't be rejecting new connections. There are some legitimate behaviors that involve multiple concurrent connections. Instead, we should keep a list of connections in that hashmap.

Does this break localhost? (all ips 127.0.0.1)

Reply:
Yes. It'd be ideal if the integration tests set up a virtual network with different IPs, but for now I guess we have a flag to turn off the limit?